docs/architecture/ADR-004-reserves.md

```diff
@@ -0,0 +1,130 @@
+
+
+## Decision
+
+The `x/leverage` module already has a module account capable of storing base asset types, and it currently does so for assets in the lending pool.
+
+Adding reserve functionality should be as simple as using the existing module account to store reserve assets, and tracking (for each stored denomination) the amount which cannot be moved from the module acount (by means of a borrower's Borrow or a lender's Withdraw) because it is reserved.
+
+The aforementioned restrictions on borrows and withdrawals must then be implemented.
+
+A governance parameter `BorrowInterestEpoch` should be used to define how many blocks to wait between interest calculations on borrow positions (e.g. "interest accrues every 100 blocks").
+Additionally, a governance parameter `ReserveFactor` must be kept which specifies the portion of borrow interest that will be funnelled into reserves. This parameter will either be decided per token, or as a single value for all asset types.
+
+> TODO: Decide on the above.
+
+Regarding the increasing of reserves when interest accrues on borrow positions, the following approach can be used (which interacts with the lender interest stream defined in [ADR-001](./ADR-001-interest-stream.md)).
+
+For each token type:
+
+- The `x/leverage` has a non-negative balance of the token in the module account
+- There is a non-negative sum of all open borrow positions using that token type, which can be calculated by the `x/leverage` module.
+- The `x/leverage` module records a non-negative amount of the token which is declared to be reserved. This amount starts at zero and increases whenever borrow interest accrues.
+
+Then:
+
+- For every open borrow of every token type, interest accrues every `BorrowInterestEpoch` (e.g. every 100 blocks) during EndBlock using a dynamically calculated interest rate.
+- At the same moment, a portion of that increase is applied to the amount of the token that is declared to be reserved.
+- The token:uToken exchange rate, which would have previously calculated by `(total tokens borrowed + module account balance) / uTokens in circulation` for a given token type and associated uToken denomination, must now be computed as `(total tokens borrowed + module account balance - reserved amount) / uTokens in circulation`.
+
+## Detailed Design
+
+The `x/leverage` module account can be used without modification to store reserves.
+
+The `x/leverage` module keeper must store, for each token denom, an `sdk.Int` representing the amount of that token which is reserved (i.e. may not be borrowed or withdrawn from the module account.)
+The key to store the reserve amount per unique denomination is formed by:
+```go
+    // reserveprefix | denom
+	var key []byte
+	key = append(key, KeyPrefixReserveAmount...)
+	key = append(key, []byte(tokenDenom)...)
+	return append(key, 0) // append 0 for null-termination
+```
```

Comment:
```suggestion
```go
    // reserveprefix | denom
    var key []byte
    key = append(key, KeyPrefixReserveAmount...)
    key = append(key, []byte(tokenDenom)...)
    return append(key, 0) // append 0 for null-termination
```
```